### PR TITLE
implementation of library uploading.

### PR DESCRIPTION
**Pull request by dakriy**
Implemented the library upload endpoints for zip file or git url.

### CODE REVIEW

**Comment by dakriy** [comment]
Closes #72

File: site/app/controllers/admin/LibraryManageController.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * @Route("/homework/library/upload/git", methods={"POST"})
+     * @return Response
+     */
+    public function ajaxUploadLibraryFromGit(): Response {
+        $useCase = new LibraryAddUseCase($this->core);
+
+        // Equivalent to $_POST['git_url'] except for it doesn't generate a notice error.
+        // Idk why E_NOTICE is enabled. It's dumb because it makes me have to do workarounds like this all over
+        $url = (isset($_POST['git_url'])) ? $_POST['git_url'] : null;
```
Comment (dakriy) [inline site/app/controllers/admin/LibraryManageController.php]: I fixed the way I do this to be way nicer in a later PR.

File: site/app/controllers/admin/LibraryManageController.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * @Route("/homework/library/upload/zip", methods={"POST"})
+     * @return Response
+     */
+    public function ajaxUploadLibraryFromZip(): Response {
+        $useCase = new LibraryAddUseCase($this->core);
+
+        // Equivalent to $_FILES['zip'] except for it doesn't generate a notice error.
+        // Idk why E_NOTICE is enabled. It's dumb because it makes me have to do workarounds like this all over
+        $file = (isset($_FILES['zip'])) ? $_FILES['zip'] : null;
```
Comment (dakriy) [inline site/app/controllers/admin/LibraryManageController.php]: I fixed the way I do this to be way nicer in a later PR.

**Comment by dakriy** [comment]
Hey @CalebJ2 thanks for the review. You are probably getting the HTML either because of the CSRF token as you mentioned or because the access control has not been finished and so for some reason the instructor user does not have the correct permissions to hit the endpoint. If you remove the @AccessControl directive at the top of the controller it will probably work.

As for the new architecture, I figure it was all right since this is a relatively large new feature, and we will need to be resuing a lot of the same code. Doing things this way allows the system to be easily testable. It becomes an intrinsic property of the system. It is also a bit different from other things they are doing because a lot of it is business logic. Currently Submitty does not really have a good place for business logic. When I asked them they told me that sort of thing can go into the library folder. I also spoke with Professor Carman about doing this, and he agrees that it is probably a good idea to pull all of the logic out of the controller for readability and maintainability. This method happens to be the best way I know of to accomplish that. Also, using this architecture, the system becomes very easy to maintain, or add new features too. It is quite scalable, very tested and I think the benefits are well worth it as it also provides framework independence as well as a clean separation between all of the other benefits previously mentioned combined with the logic, presentation, and implementation layers having a nice clean separation. If you have another idea that you think is better, I am all ears!

As for the file paths. I kind of agree, but I also feel they help convey what things are and help organize some of the classes. For example, libraries is the Submitty folder for all the extraneous code that isn't controllers or tests etc. I created the homework folder to separate this from all of the other custom library code and separate this into the homework feature. The Gateways folder is pretty obvious, it's just gateways. The next Library folder inside of that is meant to segregate those gateways off from any others that might be need to be created in the future, such as for validating a library, or pulling all of the information off of it. Although you're probably right and it may need a little work, although they should be super easy to refactor if you want. If you have any suggestions I'm happy to hear them out!

URLS! Yeah... I just did the URL's that way because that's what some of the other groups had already started doing so... Might as well try to make things at least somewhat consistent.

File: site/app/libraries/homework/Gateways/Library/FileSystemLibraryGateway.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        $sanitizedLocation = escapeshellarg($location);
+
+        $cmd = "git clone $sanitizedRepoUrl $sanitizedLocation";
+
+        $descriptors = [
+            ["pipe", "r"],  // stdin
+            ["pipe", "w"],  // stdout
+            ["pipe", "w"],  // stderr
+        ];
+        $git = proc_open($cmd, $descriptors, $pipes);
```
Comment (PlasmaIntec) [inline site/app/libraries/homework/Gateways/Library/FileSystemLibraryGateway.php]: Where exactly is `$pipes` coming from? Is this just some PHP global?

From the [Docs](https://www.php.net/manual/en/function.proc-open.php) it seems `$pipes` is just set automatically... Is this good practice?

File: site/app/libraries/homework/UseCases/LibraryAddUseCase.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        $response = new static;
+        $response->error = $message;
+        return $response;
+    }
+
+}
+
+class LibraryAddUseCase extends BaseUseCase {
+    /** @var LibraryGateway */
+    protected $gateway;
```
Comment (PlasmaIntec) [inline site/app/libraries/homework/UseCases/LibraryAddUseCase.php]: Is this permitted by [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html)? 

Does it break the Dependency Rule, "Nothing in an inner circle can know anything at all about something in an outer circle", to have a UseCase reference a Gateway?

File: site/tests/app/controllers/admin/LibraryManageControllerTester.php [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            'message' => 'A file must be provided.'
+        ], $response->json);
+        $this->assertCount(0, $this->gateway->getAllLibraries($this->location));
+    }
+
+    /** @test */
+    public function testAjaxUploadLibraryFromZipSuccess() {
+        $_FILES['zip'] = [
+            'name' => 'lib.zip',
+            'tmp_name' => 'We all make mistakes in the heat of passion, jimbo.'
```
Comment (PlasmaIntec) [inline site/tests/app/controllers/admin/LibraryManageControllerTester.php]: New Meme Invest Now!

File: site/app/libraries/homework/Gateways/Library/InMemoryLibraryGateway.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        return $this->addLibrary($location);
+    }
+
+    /** @inheritDoc */
+    public function addZipLibrary(string $filePath, string $location): string {
+        return $this->addLibrary($location);
+    }
+
+    /** @inheritDoc */
+    public function getAllLibraries(string $location): array {
```
Comment (PlasmaIntec) [inline site/app/libraries/homework/Gateways/Library/InMemoryLibraryGateway.php]: This appears to be a good place to use the [Null Coalesce Operator](https://wiki.php.net/rfc/isset_ternary)

File: site/app/libraries/homework/Gateways/LibraryGateway.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
+<?php namespace app\libraries\homework\Gateways;
+
+
+interface LibraryGateway {
+    /**
+     * This will clone the provided repo url into the specified location.
+     * On success will return 'success', and on failure will return an error message
+     *
+     * @param string $repoUrl
```
Comment (PlasmaIntec) [inline site/app/libraries/homework/Gateways/LibraryGateway.php]: Is there a design reason that on both success and failure, a string is returned? Wouldn't it be more conventional if this function returned a boolean, or a wrapped Response instance?

File: site/tests/app/controllers/admin/LibraryManageControllerTester.php [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        $this->assertEquals([
+            'status' => 'success',
+            'data' => 'Successfully cloned https://github.com/Submitty/Submitty.git.'
+        ], $response->json);
+        $this->assertCount(1, $this->gateway->getAllLibraries($this->location));
+    }
+
+    /** @test */
+    public function testAjaxUploadLibraryFromGitFail() {
+        $_POST['git_url'] = 'my first git url. BABY SHARK DO DO DO DO DO';
```
Comment (PlasmaIntec) [inline site/tests/app/controllers/admin/LibraryManageControllerTester.php]: left shark best shark 🐬

File: site/tests/app/controllers/admin/LibraryManageControllerTester.php [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+        $this->controller->ajaxUploadLibraryFromZip()->json_response;
+
+        $this->assertFileNotExists($file);
+        $this->assertTrue(FileUtils::recursiveRmdir($this->location));
+    }
+
+    protected function createFile(string $name): string {
+        $name = FileUtils::joinPaths($this->location, $name);
+        touch($name);
```
Comment (PlasmaIntec) [inline site/tests/app/controllers/admin/LibraryManageControllerTester.php]: It seems that this [line would only work if invoked from a command line](https://www.php.net/manual/en/function.touch.php). 

I suggest adding \_\_DIR\_\_ as the first argument.

File: site/tests/app/libraries/homework/UseCases/LibraryAdd/ZipLibraryAddTester.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            'tmp_name' => 'save your work boys...'
+        ]);
+
+        $this->assertEquals('A .zip file must be provided.', $this->response->error);
+    }
+
+    public function testItDoesNotAddAnInvalidFileName() {
+        $this->handleTest([
+            'name' => '/the/full/file/path/lib.zip',
+            'tmp_name' => 'I\'m slightly tired.'
```
Comment (PlasmaIntec) [inline site/tests/app/libraries/homework/UseCases/LibraryAdd/ZipLibraryAddTester.php]: 😪

File: site/tests/app/libraries/homework/UseCases/LibraryAdd/ZipLibraryAddTester.php [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            'tmp_name' => 'Thanos did nothing wrong'
+        ]);
+
+        $this->assertEquals('A file must be provided.', $this->response->error);
+    }
+
+    /** @test */
+    public function testItRequiresTmpNameAttribute() {
+        $this->handleTest([
+            'name' => 'I didn\'nt save work and lost a bunch of stuff here :( So now I\'m having to rewrite it.'
```
Comment (PlasmaIntec) [inline site/tests/app/libraries/homework/UseCases/LibraryAdd/ZipLibraryAddTester.php]: 😳

File: site/app/controllers/admin/LibraryManageController.php [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Function for adding a homework library to the server from a git repository. This should be called via AJAX,
+     * saving the result to the json_buffer of the Output object, returns a true or false on whether or not it
+     * succeeded.
+     *
+     * @Route("/homework/library/upload/git", methods={"POST"})
+     * @return Response
+     */
+    public function ajaxUploadLibraryFromGit(): Response {
```
Comment (PlasmaIntec) [inline site/app/controllers/admin/LibraryManageController.php]: For whatever reason, POST requests are being automatically redirected to `\home`, even given the changes to WebRouter.php

File: site/app/libraries/homework/Gateways/Library/FileSystemLibraryGateway.php (hunk 3 above)
Comment (dakriy) [reply site/app/libraries/homework/Gateways/Library/FileSystemLibraryGateway.php]: What? PHP globals are all caps and start with an underscore. `$pipes` is defined to be the array you see right before this. I'm just setting up descriptors for stdin, stdout, and stderr, but I only really care about stderr. I think it is just fine to be explicit about what I want. `$pipes` is an array telling PHP that I want to pipe it into memory streams that I can manipulate, and that the process should read from stdin, and write to stdout and stderr. We can also write to the programs stdin if we need to.

File: site/app/libraries/homework/UseCases/LibraryAddUseCase.php (hunk 4 above)
Comment (dakriy) [reply site/app/libraries/homework/UseCases/LibraryAddUseCase.php]: Yes, this is permitted becuase I am using [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle). Uncle bob talks about this in the link you mentioned under the section 'Crossing boundaries'.

File: site/tests/app/controllers/admin/LibraryManageControllerTester.php (hunk 5 above)
Comment (dakriy) [reply site/tests/app/controllers/admin/LibraryManageControllerTester.php]: HUZZAH! A man of quality!

File: site/app/libraries/homework/Gateways/LibraryGateway.php (hunk 7 above)
Comment (dakriy) [reply site/app/libraries/homework/Gateways/LibraryGateway.php]: On failure we want to convey why it failed, so a boolena would not be sufficient. Since we are at the gateway level, we wouldn't really be returning a response instance. It could be wrapped in a Data Transfer Object (DTO) or Entity, but I felt it was kinda overkill in this case because I don't ever see a case where we would need to return anything other than an error message. In the case of a success, the error message would be success. I suppose on success we could return the created entity... That might actually be better. What do you think?

File: site/app/libraries/homework/Gateways/Library/InMemoryLibraryGateway.php (hunk 6 above)
Comment (dakriy) [reply site/app/libraries/homework/Gateways/Library/InMemoryLibraryGateway.php]: Yeah, I guess I could. Idk if it matters too much, but I guess it's a bit nicer so, I'll do that.

File: site/tests/app/controllers/admin/LibraryManageControllerTester.php (hunk 9 above)
Comment (dakriy) [reply site/tests/app/controllers/admin/LibraryManageControllerTester.php]: Three things
 1. This is a test, when are you executing tests from the website?
 2. I don't want `__DIR__` here as the file I'm creating for testing needs to be in the special temp directory I created specifically for the test that gets deleted when the tests are done.
 3. I assume you are referring to relative file paths in which case if you look at the line right above (132) I am appending it with the full file path. So it can be run from any directory just fine without __DIR__. Perhaps I should use a different variable name to make that more clear.

File: site/app/controllers/admin/LibraryManageController.php (hunk 12 above)
Comment (dakriy) [reply site/app/controllers/admin/LibraryManageController.php]: This is because access control has not been setup. @sballance is working on that currently. In the mean time you can remove the `@accesscontrol` line in the top of the controller. It could also be that for POST requests you need to provide a CSRF token that gets handled in the WebRouter which you are not. I'm sure it's one of those two things.

File: site/tests/app/libraries/homework/UseCases/LibraryAdd/ZipLibraryAddTester.php (hunk 11 above)
Comment (dakriy) [reply site/tests/app/libraries/homework/UseCases/LibraryAdd/ZipLibraryAddTester.php]: It was the big sad.

File: site/app/libraries/homework/Gateways/LibraryGateway.php (hunk 7 above)
Comment (dakriy) [reply site/app/libraries/homework/Gateways/LibraryGateway.php]: Actually, all of the changes I need for this are in a later PR, because I think I changed the way I did some of this stuff... Sooo. Yeah.

File: site/app/libraries/homework/Gateways/Library/FileSystemLibraryGateway.php (hunk 3 above)
Comment (dakriy) [reply site/app/libraries/homework/Gateways/Library/FileSystemLibraryGateway.php]: Disregard my last comment on this. I didn't look to see exactly what you were referencing. The function proc_open fills out that variable for me. It's a pass by reference thing, and sicne it doesn't exist yet, it creates it and fills out the actual pipe descriptors for me based upon my requested descriptors that I passed in. So it opens them and then that's how it returns them, but it also needs to return the process handle, so it can't just return them. Hence the need for the pass by reference.

TL;DR:
I'm creating `$pipes` right here and proc_open is filling it out for me because it's a pass by reference.

File: site/app/libraries/homework/Gateways/LibraryGateway.php (hunk 7 above)
Comment (dakriy) [reply site/app/libraries/homework/Gateways/LibraryGateway.php]: I don't have the infrastructure in this PR, and it is created in a later one. So I'll add it in a later one where it makes more sense. This PR didn't target that as functionality. It has been done in PR #105 on commit 94c5e8e

**Comment by CalebJ2** [comment]
Tested uploading a zip file and from git successfully. Had to disable access control, send csrf_token, and enabled the homework library in config/homework_library.json.

**Comment by dakriy** [comment]
@CalebJ2 Do you want to modify the branch attached to #115 so that we can have something to collectively demo on Friday?

**Comment by CalebJ2** [comment]
Branch with my changes here: https://github.com/wwu-cs/Submitty/tree/108-109-ajax-admin-forms
